File: go/hamming/hamming.go
package hamming

import "errors"

// Distance calculates distance between two dna strands
func Distance(a, b string) (int, error) {
	ar, br := []rune(a), []rune(b)

	if len(ar) != len(br) {
		return 0, errors.New("lengths of a and b must be equal")
	}

	dist := 0
	for i := range ar {
		if ar[i] != br[i] {
			dist++
		}
	}

	return dist, nil
}

File: c/hamming/src/hamming.c
#include "hamming.h"

#include <string.h>

// calculates hamming distance between two dna strands
int compute(const char* first_strand, const char* second_strand)
{
    int distance = 0;

    if (first_strand == NULL || second_strand == NULL)
        return -1;

    if (strlen(first_strand) != strlen(second_strand))
        return -1;

    while (*first_strand && *second_strand)
    {
        if (*first_strand++ != *second_strand++)
            ++distance;
    }

    return distance;
}
